### PR TITLE
Avoid brittle flags encoding for Java enums

### LINKED ISSUE #12800: Exhaustivity checker emits spurious warning when matching on Java enum type (2.13.11 regression)
## Reproduction steps

```sh
git clone git@github.com:vasilmkd/scala-java-enum-regression.git
cd scala-java-enum-regression
sbt compile
```

https://github.com/vasilmkd/scala-java-enum-regression

Scala version: 2.13.11

## Problem (`-Werror` flag is on)
```
[error] <path>/scala-java-enum-regression/src/main/scala/main.scala:5:23: match may not be exhaustive.
[error] It would fail on the following inputs: (_ : com.intellij.openapi.compiler.CompilerMessageCategory$1), (_ : com.intellij.openapi.compiler.CompilerMessageCategory$2), (_ : com.intellij.openapi.compiler.CompilerMessageCategory$3), (_ : com.intellij.openapi.compiler.CompilerMessageCategory$4)
[error]     val messageType = category match {
[error]                       ^
[error] one error found
```

## Further context

1. Scala 2.13.10 does not exhibit this behavior.
2. Compiling using Scala 3.3.0 and Scala 2.12.18 works fine.
3. The warning/error is not issued if the same Java source code is copied into a `.java` file and compiled using `sbt`. It is necessary for this enum to be published in a library. (Might be a problem with classfile decompilation?)

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes scala/bug#12800

In the current scheme, an enum should be `sealed abstract` and have only its elements as "children".

If an enum element has a body, then permittedSubclasses includes the element subclass.

The two tweaks are: don't add those permittedSubclasses as children; set `sealed abstract` unconditionally (on parsing an enum field, where the enum is already marked sealed).

The symptoms were that patmat would warn about `MyEnum$1` (where the anon subclass is a child) or `MyEnum forSome not in X,Y,Z` (if the enum is not marked abstract).

Second commit expunges the kludge of making enum always abstract just for exhaustivity. Instead, patmat analysis checks for an enum class (`isSealedEnum`). Class file parser assumes only present flags, except `ACC_ENUM` is translated to include `SEALED` for classes. Java parser checks enum members for a deferred member, and sets `ABSTRACT` accordingly.

### CODE REVIEW

**Comment by SethTisue** [comment]
> Second commit expunges the kludge of making enum always abstract just for exhaustivity

Is that kludge something we've had since forever, or something that was added recently?

I guess I'm asking partly out of curiosity and partly because I'm wondering what the risk level is of making a code improvement that isn't strictly necessary to fix bug (right?).

In other words, my "shouldn't this be two PRs" finger is twitching slightly, but my "Seth don't be pedantic" finger is twitching in the direction of the first finger, if the metaphor isn't getting too tortured.

**Comment by SethTisue** [comment]
To overthink a little further, I guess whether I care about one-PR-or-two depends on whether 2.13.12 ends up being a longer-incubating, optimistically-merge-things type release (as 2.13.11 was), or whether it ends up being an "oh shit we need to roll another one pronto" type release (as 2.13.10 was). And we don't know that yet.

**Comment by som-snytt** [comment]
@SethTisue I expect @lrytz to weigh in on the second commit, hopefully with a ringing endorsement. Otherwise it can be moved to a separate PR as an internal improvement. (It's an old state of affairs, but I think it turns out to be low-hanging fruit. If I'd been at the recent spree, I would have worn my "low-hanging fruit" t-shirt.)

File: src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala [hunk 1]
```diff
@@ -305,24 +305,17 @@ abstract class BTypesFromSymbols[G <: Global](val global: G) extends BTypes {
...
-        // here we recover the actual classfile flags.
-        ( if (classSym.hasJavaAnnotationFlag)                        ACC_ANNOTATION | ACC_INTERFACE | ACC_ABSTRACT else 0) |
+      // scala/bug#9393: the classfile / java source parser make java annotation symbols look like classes.
+      // here we recover the actual classfile flags.
+      ( if (classSym.hasJavaAnnotationFlag)                        ACC_ANNOTATION | ACC_INTERFACE | ACC_ABSTRACT else 0) |
       ( if (classSym.isPublic)                                     ACC_PUBLIC    else 0) |
       ( if (classSym.isFinal)                                      ACC_FINAL     else 0) |
        // see the link above. javac does the same: ACC_SUPER for all classes, but not interfaces.)
       ( if (classSym.isInterface)                                  ACC_INTERFACE else ACC_SUPER) |
        // for Java enums, we cannot trust `hasAbstractFlag` (see comment in enumFlags))
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala]:
```suggestion
```

File: src/reflect/scala/reflect/internal/ClassfileConstants.scala [hunk 2]
```diff
@@ -353,7 +353,7 @@ object ClassfileConstants {
       case JAVA_ACC_STATIC     => STATIC
       case JAVA_ACC_ABSTRACT   => if (isClass) ABSTRACT else DEFERRED
       case JAVA_ACC_INTERFACE  => TRAIT | INTERFACE | ABSTRACT
-      case JAVA_ACC_ENUM       => JAVA_ENUM
+      case JAVA_ACC_ENUM       => if (isClass) JAVA_ENUM | SEALED else JAVA_ENUM
```
Comment (lrytz) [inline src/reflect/scala/reflect/internal/ClassfileConstants.scala:356]: Here I don't follow, when is `isClass` true, when is it false?

**Review by lrytz — COMMENTED** [review]
LGTM, both commits, just the `isSealedEnum` extension method seems a bit of an overkill.

File: src/reflect/scala/reflect/internal/ClassfileConstants.scala (hunk 2 above)
Comment (som-snytt) [reply src/reflect/scala/reflect/internal/ClassfileConstants.scala:356]: The enum member fields get the flag. (I believe I did not imagine that.)

File: src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala (hunk 1 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/backend/jvm/BTypesFromSymbols.scala]: Thanks, I forgot to return to this spot. See the funky formatting of the next line.

**Comment by som-snytt** [comment]
Updated with two suggestions. (Edit: and rebased)

In the beginning, I expected to be calling `isBlahBlah` all over the place, but I see it wound up as just a tweak to two expressions.

That is not in proportion to the degree of suffering required.

Also adjusted indentation on the flag translation, so that it fits in my default terminal.

Reluctantly renamed variable `abbie` to `hasAbstractFlag` (change not explicitly requested).

I did not squash the two commits, but I did squash the two amendments. Dangerous to do before finishing first coffee! After deleting the package object, it tried to leak back into the second commit. ("How does git actually work again?")

**Review by lrytz — APPROVED** [review]
> Reluctantly renamed variable `abbie` to `hasAbstractFlag` (change not explicitly requested).

I like sentimental codebases

**Pull request merged**